### PR TITLE
fix: refactor triggerNextJobs function

### PR DESCRIPTION
**Pull request by jithine**
## Context

`triggerNextJobs` function is responsible for running next jobs after current running job is success. It has grown too big and code has become difficult to track.

## Objective

1. Remove externalJoin flag and make remote-join feature permanent.
2. refactor code to clear flows

## References

https://github.com/screwdriver-cd/screwdriver/pull/2374#issuecomment-812766411

## License
I confirm that this contribution is made under a BSD license and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: plugins/builds/index.js [hunk 1]
```diff
@@ -950,53 +950,52 @@ async function createOrRunNextBuild({
...
         const [fullName, pId, jName] = EXTERNAL_TRIGGER_ALL.exec(n);
 
+        externalJobPipelineIds.push(pId);
+
         if (externalTriggersDict[pId]) {
-            externalTriggersDict[pId].full = externalTriggersDict[pId].full.concat([fullName]);
-            externalTriggersDict[pId].short = externalTriggersDict[pId].short.concat([jName]);
+            duplicatePipelineIds.push(pId);
+            externalTriggersDict[pId].full.push(fullName);
+            externalTriggersDict[pId].push(jName);
```
Comment (tkyi) [inline plugins/builds/index.js:998]:
```suggestion
            externalTriggersDict[pId].short.push(jName);
```

File: plugins/builds/index.js [hunk 2]
```diff
@@ -1154,8 +1202,12 @@ const buildsPlugin = {
                 trigger: currentJobName,
                 chainPR: pipeline.chainPR
             });
-            // Create a join object like: {A:[B,C], D:[B,F]} where [B,C] join on A, [B,F] join on D, etc.
-            // This can include external jobs
+            /*
+             * Create a join object :
```
Comment (tkyi) [inline plugins/builds/index.js:1206]:
```suggestion
             * Create a join object:
```

File: plugins/builds/index.js [hunk 3]
```diff
@@ -1101,6 +1097,58 @@ async function handleDuplicatePipelines(config) {
...
+                username,
+                scmContext,
+                build, // this is the parentBuild for the next build
+                baseBranch: event.baseBranch || null
+            };
+
+            // Just start the build if falls in to these 2 scenarios
+            // 1. No join
+            // 2. ([~D,B,C]->A) currentJob=D, nextJob=A, joinList(A)=[B,C]
+            //    joinList doesn't include C, so start A
```
Comment (tkyi) [inline plugins/builds/index.js:1133]:
```suggestion
            // Just start the build if falls into these 2 scenarios
            // 1. No join
            // 2. ([~D,B,C]->A) currentJob=D, nextJob=A, joinList(A)=[B,C]
            //    joinList doesn't include D, so start A
```

File: plugins/builds/index.js [hunk 4]
```diff
@@ -1101,6 +1097,58 @@ async function handleDuplicatePipelines(config) {
     return joinObj;
 }
 
+/**
+ * Get pipelineId and job name from the `name`
+ * If internal, pipelineId will be the current pipelineId
+ * @param  {String} name        Job name
+ * @param  {String} pipelineId  Pipeline ID
+ * @return {Object}             With pipeline id and job name
```
Comment (tkyi) [inline plugins/builds/index.js:1105]: need to fix jsdocs

File: plugins/builds/index.js [hunk 5]
```diff
@@ -1166,49 +1218,14 @@ const buildsPlugin = {
...
-                );
+                return triggerNextJobsWithNoExternalJoin(joinObj, event, app, config);
             }
 
             /* NEW FLOW
              * Use if external join flag is true
              */
-            // Trigger jobs with duplicate pipelines first; remove them from joinObj
-            joinObj = await handleDuplicatePipelines({
+            // Trigger jobs with same pipeline first; remove them from joinObj
```
Comment (tkyi) [inline plugins/builds/index.js:1227]:
```suggestion
            // Trigger external jobs with same pipeline and no join first; remove them from joinObj
```

File: plugins/builds/index.js [hunk 6]
```diff
@@ -757,347 +545,80 @@ async function getParallelBuilds({ eventFactory, parentEventId, pipelineId }) {
...
- * Finds unique pipeline IDs and filters them out to return duplicates
- * @param  {Array} externalJobPipelineIds External job pipeline IDs
- * @return {Array}                        Duplicate external job pipeline IDs
+ * Create joinObject for nextJobs to trigger
+ *   For A & D in nextJobs for currentJobName B, create
+ *          {A:[B,C], D:[B,F], X: []} where [B,C] join on A,
+ *              [B,F] join on D and X has no join
+ *   This can include external jobs
+ * @param {Array}  nextJobs
+ * @param {Object} event
```
Comment (tkyi) [inline plugins/builds/index.js:588]: need to fix docs

File: bin/server [hunk 7]
```diff
@@ -58,8 +58,6 @@ Object.keys(clusterEnv).forEach(k => {
     clusterEnv[k] = String(clusterEnv[k]);
 });
 
-// Flag to allow external join
-const { externalJoin } = config.get('build');
```
Comment (tkyi) [inline bin/server:62]: can you remove from config files as well?
https://github.com/screwdriver-cd/screwdriver/blob/new/config/custom-environment-variables.yaml#L406
https://github.com/screwdriver-cd/screwdriver/blob/new/config/default.yaml#L320

File: plugins/builds/index.js [hunk 8]
```diff
@@ -757,345 +546,84 @@ async function getParallelBuilds({ eventFactory, parentEventId, pipelineId }) {
...
- * @param  {Object}     parentBuilds        Parent builds info
- * @param  {Number}     parentEventId       Parent event ID
- * @param  {Number}     parentBuildId       Parent build ID
- * @param  {Boolean}    isExternal          Is external or not
- * @param  {Build}      externalBuild       External build
- * @param  {Array}      joinListNames       Join list names
- * @param  {Object}     currentJobParentBuilds Parent builds info for current job
- * @param  {Object}     currentBuildInfo    Parent builds info for current build
- * @return {Promise}                        The newly updated/created build
+ * Fills parentBuilds object with missing job information
```
Comment (tkyi) [inline plugins/builds/index.js:549]: need to add JSdocs

File: plugins/builds/index.js [hunk 9]
```diff
@@ -757,345 +551,88 @@ async function getParallelBuilds({ eventFactory, parentEventId, pipelineId }) {
...
-    const { joinObj, pipelineFactory, eventFactory, pipelineId, currentJobName, build, event } = config;
-    const { duplicatePipelineIds, externalTriggersDict } = parseJoinObj(joinObj);
-    const pipelinesToStart = [];
-
-    // Get pipeline's workflowGraph to make sure there is no join
-    await Promise.all(
-        duplicatePipelineIds.map(async id => {
-            const duplicateJobNames = externalTriggersDict[id].short;
-            const pipeline = await pipelineFactory.get(id);
+        if (pid !== 'current') {
```
Comment (pritamstyz4ever) [inline plugins/builds/index.js:621]: may be we can convert this to constant

**Review by pritamstyz4ever — APPROVED** [review]

File: plugins/builds/index.js [hunk 10]
```diff
@@ -757,345 +551,88 @@ async function getParallelBuilds({ eventFactory, parentEventId, pipelineId }) {
...
- * @param  {Boolean}    isExternal          Is external or not
- * @param  {Build}      externalBuild       External build
- * @param  {Array}      joinListNames       Join list names
- * @param  {Object}     currentJobParentBuilds Parent builds info for current job
- * @param  {Object}     currentBuildInfo    Parent builds info for current build
- * @return {Promise}                        The newly updated/created build
+ * Fills parentBuilds object with missing job information
+ * @param {Array}  parentBuilds
+ * @param {Object} current       Holds current build/event data
+ * @param {builds} builds        Completed builds which is used to fill parentBuilds data
```
Comment (tkyi) [inline plugins/builds/index.js:557]:
```suggestion
 * @param {Array} builds        Completed builds which is used to fill parentBuilds data
```

File: plugins/builds/index.js [hunk 11]
```diff
@@ -757,345 +551,88 @@ async function getParallelBuilds({ eventFactory, parentEventId, pipelineId }) {
...
- *   }
- * }
- * @param  {Object} joinObj Join object
- * @return {Object}         Duplicate pipeline IDs and external triggers dict
+ * Create joinObject for nextJobs to trigger
+ *   For A & D in nextJobs for currentJobName B, create
+ *          {A:[B,C], D:[B,F], X: []} where [B,C] join on A,
+ *              [B,F] join on D and X has no join
+ *   This can include external jobs
+ * @param {Array}   nextJobs
```
Comment (tkyi) [inline plugins/builds/index.js:597]: missing description

File: test/plugins/builds.test.js [hunk 12]
```diff
@@ -1787,29 +1610,38 @@ describe('build plugin test', () => {
...
+                    };
+                    const parentBuildsC = {
+                        123: { eventId: '8888', jobs: { a: 12345, d: 123456 } }
+                    };
+
+                    buildFactoryMock.create.onCall(0).returns({ ...buildMock, parentBuilds: parentBuildsB });
+                    buildFactoryMock.create.onCall(1).returns({ ...buildMock, parentBuilds: parentBuildsC });
+
+                    // buildFactoryMock.get.withArgs(12345).resolves(buildMocks[0])
+                    // buildFactoryMock.get.withArgs(123456).resolves(buildMocks[1]);
```
Comment (tkyi) [inline test/plugins/builds.test.js:1641]: commented code

File: test/plugins/builds.test.js [hunk 13]
```diff
@@ -3523,25 +3193,25 @@ describe('build plugin test', () => {
...
                     const eventConfig = {
                         causeMessage: 'Triggered by sd@123:a',
-                        groupEventId: '8888',
+                        groupEventId: '8889',
                         parentBuildId: 12345,
                         parentBuilds: {
-                            123: { eventId: '8888', jobs: { a: 12345, c: 45678 } },
-                            2: { eventId: '8887', jobs: { a: 12345 } }
+                            123: { eventId: '8888', jobs: { a: 12345, c: 45678 } }
+                            // 2: { eventId: '8887', jobs: { a: 12345 } }
```
Comment (tkyi) [inline test/plugins/builds.test.js:3200]: commented code

File: test/plugins/builds.test.js [hunk 14]
```diff
@@ -3470,9 +3139,10 @@ describe('build plugin test', () => {
                     // ~sd@2:a -> a -> ~sd@2:c
                     //
                     // 2. pipeline 2
-                    //   ->    b     ->
+                    //   ------------->
                     // a                c
                     //   -> sd@123:a ->
+                    //                  d
                     //
                     // If user restarts `a`, it should get `d`'s parent event status and trigger `c`
```
Comment (tkyi) [inline test/plugins/builds.test.js:3147]: is this comment still accurate?

File: plugins/builds/index.js (hunk 9 above)
Comment (jithine) [reply plugins/builds/index.js]: updated

**Comment by jithine** [comment]
Testing done

Pipeline 1 https://github.com/jithine/-workflow-parallel-join-remote-example-/blob/main/screwdriver.yaml
Pipeline 2 https://github.com/jithine/workflow-parallel-join-example/blob/master/screwdriver.yaml


https://user-images.githubusercontent.com/193126/113463444-821ca100-93da-11eb-9800-f9a4cd756780.mov

File: plugins/builds/index.js [hunk 15]
```diff
@@ -533,47 +346,41 @@ function createParentBuildsObj(config) {
  * - parentBuilds: parent build information
  * - joinListNames: array of join jobs
  * - joinParentBuilds: parent build information for join jobs
- * - currentJobParentBuilds: parent build information for current job
- * - currentBuildInfo: build information for current job
  * @param  {Object} joinObj        Join object
  * @param  {String} currentJobName Current job name
```
Comment (tkyi) [inline plugins/builds/index.js:350]: need to update docs

File: plugins/builds/index.js [hunk 16]
```diff
@@ -586,54 +393,29 @@ function parseJobInfo({ joinObj, currentJobName, nextJobName, pipelineId, build
  * @param  {Factory}    [buildFactory]          Build factory
  * @return {Promise}                            All finished builds
  */
-async function getFinishedBuilds(event, eventFactory, buildFactory) {
+async function getFinishedBuilds(event, buildFactory) {
```
Comment (tkyi) [inline plugins/builds/index.js:396]: need to update docs

File: plugins/builds/index.js [hunk 17]
```diff
@@ -586,54 +393,29 @@ function parseJobInfo({ joinObj, currentJobName, nextJobName, pipelineId, build
...
 /**
  * Update parent builds info when next build already exists
  * @param  {Object} joinParentBuilds       Parent builds object for join job
- * @param  {Object} currentJobParentBuilds Parent builds object for current job
  * @param  {Build}  nextBuild              Next build
- * @param  {Object} currentBuildInfo       Build info for current job
  * @return {Promise}                       Updated next build
  */
-async function updateParentBuilds({ joinParentBuilds, currentJobParentBuilds, nextBuild, currentBuildInfo, build }) {
+async function updateParentBuilds({ joinParentBuilds, nextBuild, build }) {
```
Comment (tkyi) [inline plugins/builds/index.js:411]: need to update docs

File: plugins/builds/index.js [hunk 18]
```diff
@@ -46,159 +46,20 @@ function getPipelineAndJob(name, pipelineId) {
...
- * @param  {Number}   config.pipelineId     Pipeline Id
- * @param  {String}   config.jobName        Job name
- * @param  {String}   config.username       Username of build
- * @param  {String}   config.scmContext     SCM context
- * @param  {Build}    config.build          Build object
- * @param  {Boolean}  [config.start]        Whether to start the build or not
- * @param  {String}   config.baseBranch     Branch name
- * @return {Promise}
+ * Helper function to fetch external event from parentBuilds
+ * @param  {Object} currentBuild
```
Comment (tkyi) [inline plugins/builds/index.js:50]: missing descriptions

File: plugins/builds/index.js [hunk 19]
```diff
@@ -46,159 +46,20 @@ function getPipelineAndJob(name, pipelineId) {
...
+ * @param  {String} pipelineId
+ * @param  {Object} eventFactory
+ * @return {Object} External Event
```
Comment (tkyi) [inline plugins/builds/index.js:53]: missing descriptions

File: plugins/builds/index.js [hunk 20]
```diff
@@ -312,7 +173,8 @@ async function createExternalBuild(config) {
         parentBuilds,
         causeMessage,
         parentEventId,
-        groupEventId
+        groupEventId,
+        start
```
Comment (tkyi) [inline plugins/builds/index.js:177]: need to update jsdocs for `start`

File: plugins/builds/index.js (hunk 16 above)
Comment (jithine) [reply plugins/builds/index.js:389]: done

File: plugins/builds/index.js (hunk 17 above)
Comment (jithine) [reply plugins/builds/index.js:405]: done

File: plugins/builds/index.js [hunk 21]
```diff
@@ -1139,360 +665,291 @@ const buildsPlugin = {
...
                  */
-                if (joinListNames.length === 0 || currentJobNotInJoinList) {
-                    // Next build is internal
-                    if (!isExternal) {
-                        const internalBuildConfig = {
-                            jobFactory,
-                            buildFactory,
+                const isORTrigger = !joinListNames.includes(current.job.name);
+
+                if (joinListNames === 0 || isORTrigger) {
```
Comment (tkyi) [inline plugins/builds/index.js:707]:
```suggestion
                if (joinListNames.length === 0 || isORTrigger) {
```

File: plugins/builds/index.js [hunk 22]
```diff
@@ -362,21 +223,17 @@ async function createInternalBuild(config) {
     const {
         jobFactory,
         buildFactory,
-        eventFactory,
         pipelineId,
         jobName,
         username,
         scmContext,
-        build,
+        event,
```
Comment (tkyi) [inline plugins/builds/index.js:230]: need to update jsdocs (remove build, eventFactory, eventId, sha; add event)

File: plugins/builds/index.js [hunk 23]
```diff
@@ -757,345 +532,96 @@ async function getParallelBuilds({ eventFactory, parentEventId, pipelineId }) {
...
- * @param  {Build}      externalBuild       External build
- * @param  {Array}      joinListNames       Join list names
- * @param  {Object}     currentJobParentBuilds Parent builds info for current job
- * @param  {Object}     currentBuildInfo    Parent builds info for current build
- * @return {Promise}                        The newly updated/created build
+ * Fills parentBuilds object with missing job information
+ * @param {Array}  parentBuilds
+ * @param {Object} current       Holds current build/event data
+ * @param {Array}  builds        Completed builds which is used to fill parentBuilds data
+ * @param {Object} nextEvent     External event
```
Comment (tkyi) [inline plugins/builds/index.js:539]:
```suggestion
 * @param {Object} [nextEvent]   External event
```

File: plugins/builds/index.js [hunk 24]
```diff
@@ -757,345 +532,96 @@ async function getParallelBuilds({ eventFactory, parentEventId, pipelineId }) {
...
+ * @return {Array} Array of parentBuilds with missing data filled in.
```
Comment (tkyi) [inline plugins/builds/index.js:540]:
```suggestion
```

File: plugins/builds/index.js [hunk 25]
```diff
@@ -1139,360 +665,291 @@ const buildsPlugin = {
...
+                        eventFactory,
+                        parentEventId: current.event.parentEventId,
+                        pipelineId: current.pipeline.id
+                    });
+
+                    finishedInternalBuilds = finishedInternalBuilds.concat(parallelBuilds);
+                }
+                fillParentBuilds(parentBuilds, current, finishedInternalBuilds);
+                // If next build is internal, look at the finished builds for this event
+                const jobId = joinObj[nextJobName].id;
```
Comment (tkyi) [inline plugins/builds/index.js:742]: is this next job's id?
```suggestion
                const nextJobId = joinObj[nextJobName].id;
```

File: plugins/builds/index.js [hunk 26]
```diff
@@ -1139,360 +665,291 @@ const buildsPlugin = {
...
+                const nextJobs = joinObj.jobs;
+                let nextJobNames = Object.keys(nextJobs);
+                const triggerName = `sd@${current.pipeline.id}:${current.job.name}`;
+
+                if (externalEvent) {
+                    // Remote join case
+                    // fetch builds created due to restart
+                    const externalGroupBuilds = await getFinishedBuilds(externalEvent, buildFactory);
+
+                    const buildsToRestart = Object.keys(nextJobs)
```
Comment (tkyi) [inline plugins/builds/index.js:806]:
```suggestion
                    const buildsToRestart = nextJobNames
```

**Review by tkyi — APPROVED** [review]

**Review by tkyi — APPROVED** [review]

**Pull request merged**